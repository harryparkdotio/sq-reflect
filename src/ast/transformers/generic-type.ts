import * as ts from 'typescript';
import { isGenericType } from '../generics/types';

export const genericTypeTransformer = <T extends ts.Node>(
  typeParameters: ts.TypeParameterDeclaration[]
): ts.TransformerFactory<T> => {
  return (context: ts.TransformationContext) => {
    const visit: ts.Visitor = (node: ts.Node) => {
      /* istanbul ignore else */
      if (ts.isTypeNode(node)) {
        if (ts.isArrayTypeNode(node)) {
          return ts.visitEachChild(node, child => visit(child), context);
        } else if (
          ts.isUnionTypeNode(node) ||
          ts.isIntersectionTypeNode(node)
        ) {
          return ts.createUnionOrIntersectionTypeNode(
            node.kind,
            ts.visitNodes(node.types, visit)
          );
        }

        if (isGenericType(node)) {
          // FIXME: this won't work if there is more than one type parameter
          // ideally this would have excel style lettering (A --> Z, AA --> AZ, ...)
          const typeParameterIdentifier = ts.createIdentifier('T');

          typeParameters.push(
            ts.createTypeParameterDeclaration(
              typeParameterIdentifier,
              undefined,
              node
            )
          );

          return typeParameterIdentifier;
        }

        return node;
      }

      /* istanbul ignore next */
      return ts.visitEachChild(node, child => visit(child), context);
    };

    return node => ts.visitNode(node, visit);
  };
};
