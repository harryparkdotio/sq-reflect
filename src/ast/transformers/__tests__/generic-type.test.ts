import * as ts from 'typescript';
import { genericTypeTransformer } from '../generic-type';

const transform = <T extends ts.Node>(
  node: T | T[],
  transformer: ts.TransformerFactory<T>
): T[] => ts.transform(node, [transformer]).transformed;

describe('genericTypeTransformer', () => {
  test('should transform generic types from primitive to generic', () => {
    const node = ts.createArrayTypeNode(
      ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword)
    );

    const typeParameterReferenceArray: ts.TypeParameterDeclaration[] = [];
    const transformer = genericTypeTransformer(typeParameterReferenceArray);

    const result = transform(node, transformer);

    expect(result).toMatchSnapshot();
    expect(typeParameterReferenceArray).toMatchSnapshot();
  });

  test('should transform generic types from primitive to generic', () => {
    const node = ts.createUnionTypeNode([
      ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword),
      ts.createKeywordTypeNode(ts.SyntaxKind.ObjectKeyword),
    ]);

    const typeParameterReferenceArray: ts.TypeParameterDeclaration[] = [];
    const transformer = genericTypeTransformer(typeParameterReferenceArray);

    const result = transform(node, transformer);

    expect(result).toMatchSnapshot();
    expect(typeParameterReferenceArray).toMatchSnapshot();
  });

  test('should transform generic types from primitive to generic', () => {
    const node = ts.createIntersectionTypeNode([
      ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword),
      ts.createKeywordTypeNode(ts.SyntaxKind.ObjectKeyword),
    ]);

    const typeParameterReferenceArray: ts.TypeParameterDeclaration[] = [];
    const transformer = genericTypeTransformer(typeParameterReferenceArray);

    const result = transform(node, transformer);

    expect(result).toMatchSnapshot();
    expect(typeParameterReferenceArray).toMatchSnapshot();
  });

  test('should transform generic types from primitive to generic', () => {
    const node = ts.createKeywordTypeNode(ts.SyntaxKind.StringKeyword);

    const typeParameterReferenceArray: ts.TypeParameterDeclaration[] = [];
    const transformer = genericTypeTransformer(typeParameterReferenceArray);

    const result = transform(node, transformer);

    expect(result).toMatchSnapshot();
    expect(typeParameterReferenceArray).toMatchSnapshot();
  });

  test('should not transform generic types in type literal from primitive to generic', () => {
    const node = ts.createTypeLiteralNode([
      ts.createPropertySignature(
        undefined,
        ts.createIdentifier('a'),
        undefined,
        ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword),
        undefined
      ),
      ts.createPropertySignature(
        undefined,
        ts.createIdentifier('b'),
        undefined,
        ts.createKeywordTypeNode(ts.SyntaxKind.StringKeyword),
        undefined
      ),
    ]);

    const typeParameterReferenceArray: ts.TypeParameterDeclaration[] = [];
    const transformer = genericTypeTransformer(typeParameterReferenceArray);

    const result = transform(node, transformer);

    expect(result).toMatchSnapshot();
    expect(typeParameterReferenceArray).toMatchSnapshot();
  });

  test('should', () => {
    const node = ts.createTypeReferenceNode(ts.createIdentifier('Array'), [
      ts.createTypeReferenceNode(ts.createIdentifier('Record'), [
        ts.createLiteralTypeNode(ts.createNumericLiteral('7')),
        ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword),
      ]),
    ]);

    const typeParameterReferenceArray: ts.TypeParameterDeclaration[] = [];
    const transformer = genericTypeTransformer(typeParameterReferenceArray);

    const result = transform(node, transformer);

    expect(result).toMatchSnapshot();
    expect(typeParameterReferenceArray).toMatchSnapshot();
  });
});
