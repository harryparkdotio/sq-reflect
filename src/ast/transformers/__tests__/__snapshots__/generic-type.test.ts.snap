// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`genericTypeTransformer should 1`] = `
Array [
  Array<Record<7, any>>,
]
`;

exports[`genericTypeTransformer should 2`] = `Array []`;

exports[`genericTypeTransformer should not transform generic types in type literal from primitive to generic 1`] = `
Array [
  {
    a: any;
    b: string;
},
]
`;

exports[`genericTypeTransformer should not transform generic types in type literal from primitive to generic 2`] = `Array []`;

exports[`genericTypeTransformer should transform generic types from primitive to generic 1`] = `
Array [
  TestType[],
]
`;

exports[`genericTypeTransformer should transform generic types from primitive to generic 2`] = `
Array [
  TestType = any,
]
`;

exports[`genericTypeTransformer should transform generic types from primitive to generic 3`] = `
Array [
  TestType | TestType,
]
`;

exports[`genericTypeTransformer should transform generic types from primitive to generic 4`] = `
Array [
  TestType = any,
  TestType = object,
]
`;

exports[`genericTypeTransformer should transform generic types from primitive to generic 5`] = `
Array [
  TestType & TestType,
]
`;

exports[`genericTypeTransformer should transform generic types from primitive to generic 6`] = `
Array [
  TestType = any,
  TestType = object,
]
`;

exports[`genericTypeTransformer should transform generic types from primitive to generic 7`] = `
Array [
  string,
]
`;

exports[`genericTypeTransformer should transform generic types from primitive to generic 8`] = `Array []`;
